mlserver/batch_processing.py

```diff
@@ -0,0 +1,283 @@
+        data = output_item.item.as_numpy(output["name"])
+        response["outputs"][n]["data"] = data.flatten().tolist()
+
+        # Removing "binary_data_size" parameters as now this is 1-D list
+        if (
+            output["parameters"] is not None
+            and "binary_data_size" in output["parameters"]
+        ):
+            del response["outputs"][n]["parameters"]["binary_data_size"]
+
+    return orjson.dumps(response)
+
+
+async def process_item(
+    item: BatchInputItem,
+    model_name: str,
+    worker_id: int,
+    triton_client: httpclient.InferenceServerClient,
+    binary_data: bool,
+) -> BatchOutputItem:
+    try:
+        request_id, inputs, outputs = json_to_triton(item.item, binary_data)
+    except Exception as e:
+        logger.error(f"consumer {worker_id}: failed to deserialize item: {e}")
+        raise
+    try:
+        logger.debug(f"consumer {worker_id}: sending request")
+        data = await triton_client.infer(
+            model_name,
+            inputs,
+            outputs=outputs,
+            request_id=request_id,
+            headers=get_headers(request_id),
+        )
+        logger.debug(f"consumer {worker_id}: received response")
+    except Exception as e:
+        logger.error(f"Consumer {worker_id}: failed to process task: {e}")
+        raise
+    return BatchOutputItem(index=item.index, item=data)
+
+
+async def produce(queue: asyncio.Queue, fname: str):
+    async with aiofiles.open(fname, "rb") as f:
+        index = 0
+        async for line in f:
+            await queue.put(BatchInputItem(index=index, item=line))
+            index += 1
+
+
+async def consume(
+    model_name: str,
+    worker_id: int,
+    triton_client: httpclient.InferenceServerClient,
+    binary_data: bool,
+    extra_verbose: bool,
+    queue_in: asyncio.Queue,
+    queue_out: asyncio.Queue,
+):
+    while True:
+        input_item = await queue_in.get()
+        try:
+            output_item = await process_item(
+                input_item, model_name, worker_id, triton_client, binary_data
+            )
+        except Exception:
+            if extra_verbose:
+                logger.error(
+                    f"consumer {worker_id}: failed to process item {input_item}"
+                )
+            queue_in.task_done()
+            continue
+
+        await queue_out.put(output_item)
+        queue_in.task_done()
+
+
+async def finalize(queue: asyncio.Queue, fname: str):
+    # TODO: Test if output directory is writtable, sysexit otherwise.
+    async with aiofiles.open(fname, "wb") as f:
+        while True:
+            item = await queue.get()
+            try:
+                output = serialize_triton_infer_result(item)
+                await f.write(output)
+                await f.write(b"\n")
+            except Exception as e:
+                logger.error(f"Failed to finalize task: {e}")
+            queue.task_done()
+
+
+async def process_batch(
```

Comment:
One for the design doc but there may be some value for providing a `process_batch_sync` function to simplify interaction if someone wants to call the function from a non-asyncio loop in python - otherwise they'll have to handle the asyncio logic themselves, but perhaps not as bad, maybe it could just be documentation that explains how people could do it if needed

Reply:
Actually, for now this is meant mostly as the CLI entrypoint, not to be yet used directly anywhere else. 
I am going to add the docstring.

Reply:
I added small docstrings just now

Reply:
Exposing it as `from mlserver import ...` functionality is definitely a worth following up on though :+1: